nixos/tests/k3s/default.nix

```diff
@@ -1,9 +1,22 @@
 { system ? builtins.currentSystem
 , pkgs ? import ../../.. { inherit system; }
+, lib ? pkgs.lib
 }:
-{
-  # Run a single node k3s cluster and verify a pod can run
-  single-node = import ./single-node.nix { inherit system pkgs; };
-  # Run a multi-node k3s cluster and verify pod networking works across nodes
-  multi-node = import ./multi-node.nix { inherit system pkgs; };
-}
+let
+  allK3s = {
+    inherit (pkgs)
+      k3s
+      k3s_1_23
+      k3s_1_24
+      k3s_1_25
+      k3s_1_26
+    ;
+  };
+  oneTest = k3s: {
+    # Run a single node k3s cluster and verify a pod can run
+    single-node = import ./single-node.nix { inherit system pkgs k3s; };
+    # Run a multi-node k3s cluster and verify pod networking works across nodes
+    multi-node = import ./multi-node.nix { inherit system pkgs k3s; };
+  };
```

Comment:
Not too sure about this structure. This way we get `nixosTests.k3s = { k3s_1_23 = { single-mode = <drv>; multi-mode = <drv>; }; ... };`

I think it'd be better to have `nixosTests.k3s-single-mode = { k3s_1_23 = <drv>; };` and `nixosTests.k3s-multi-mode = { k3s_1_23 = <drv>; };`. This can be instantiated more easily.

Reply:
Shouldn't it be `nixosTests.k3s.single-mode` and `nixosTests.k3s.multi-mode` instead? I think having all under the `k3s` top key is more helpful IMHO.

Reply:
That's even better but I'm not sure how hydra interacts with tests; it might require top-level attrs.

Reply:
AFAICS it should be OK:

https://github.com/NixOS/nixpkgs/blob/e77913a680bc9e7bfd2c93047ecfd6697e2402d7/nixos/tests/all-tests.nix#L20-L34

https://github.com/NixOS/nixpkgs/blob/e77913a680bc9e7bfd2c93047ecfd6697e2402d7/nixos/tests/all-tests.nix#L299

However, still worried about https://github.com/NixOS/nixpkgs/pull/215482#issuecomment-1424222311.

Is there any way to actually check that the test is being executed?

Reply:
Fixed now:

```nix
nix-repl> pkgs = import ./. {}           

nix-repl> pkgs.nixosTests.k3s.multi-node
{ k3s = «derivation /nix/store/gcqm103sppa8wqbvg0dlamc6hbpcd97z-vm-test-run-k3s-1.26.1+k3s1-multi-node.drv»; k3s_1_23 = «derivation /nix/store/y1qnffj0vsrvqf5b89n6grsf51zhd35s-vm-test-run-k3s-1.23.6+k3s1-multi-node.drv»; k3s_1_24 = «derivation /nix/store/ykkly237zy9kzm5vbsnaha0mvq13kx5f-vm-test-run-k3s-1.24.4+k3s1-multi-node.drv»; k3s_1_25 = «derivation /nix/store/ykqjdwwc1qhmpi1aj1gxkh2488b7mlm7-vm-test-run-k3s-1.25.3+k3s1-multi-node.drv»; k3s_1_26 = «derivation /nix/store/gcqm103sppa8wqbvg0dlamc6hbpcd97z-vm-test-run-k3s-1.26.1+k3s1-multi-node.drv»; }

nix-repl> pkgs.nixosTests.k3s.single-node
{ k3s = «derivation /nix/store/j9xazs9bdkmjsvf1zpbvdxm6fsf4c3hn-vm-test-run-k3s-1.26.1+k3s1-single-node.drv»; k3s_1_23 = «derivation /nix/store/d947fhzjw9kric2ggk7d3bns894d9d5y-vm-test-run-k3s-1.23.6+k3s1-single-node.drv»; k3s_1_24 = «derivation /nix/store/yip0rgij0hh6ryy4rsfnf49kdc9qj4pv-vm-test-run-k3s-1.24.4+k3s1-single-node.drv»; k3s_1_25 = «derivation /nix/store/k164z1v0z5bab5ga7a7cd74sqxjzkhxl-vm-test-run-k3s-1.25.3+k3s1-single-node.drv»; k3s_1_26 = «derivation /nix/store/j9xazs9bdkmjsvf1zpbvdxm6fsf4c3hn-vm-test-run-k3s-1.26.1+k3s1-single-node.drv»; }
```

Reply:
https://github.com/NixOS/nixpkgs/pull/215482#issuecomment-1424222311 was due to this nesting. Works fine now that the attrsets only contain derivations and aren't nested anymore. I.e. `nix-build -A nixosTests.k3s.multi-node`.